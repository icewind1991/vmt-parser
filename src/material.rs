use crate::error::{EofError, ParseValueError};
use crate::Result;
use std::collections::HashMap;
use vdf_reader::entry::{Entry, FromEntry, Table};
use vdf_reader::{Event, GroupStartEvent, Reader, VdfError};

#[derive(Debug, Clone)]
pub enum Material {
    Other(OtherMaterial),
    LightMappedGeneric(LightMappedGenericMaterial),
}

impl Material {
    pub fn parse(reader: &mut Reader) -> Result<Self> {
        let start: GroupStartEvent = expect_event(reader, "material name")?;
        let name = start.name.to_ascii_lowercase();

        Ok(match name.as_str() {
            "lightmappedgeneric" => {
                Material::LightMappedGeneric(LightMappedGenericMaterial::parse(reader)?)
            }
            _ => Material::Other(OtherMaterial::parse(name, reader)?),
        })
    }
}

#[derive(Debug, Clone)]
pub struct LightMappedGenericMaterial {
    pub keywords: String,
    pub detail: Option<String>,
    pub detail_blend_factory: f32,
    pub detail_scale: f32,
    pub detail_blend_mode: u32,
    pub base_texture: String,
    pub ss_bump: bool,
    pub bump_map: Option<String>,
    pub rest: HashMap<String, Entry>,
}

impl LightMappedGenericMaterial {
    fn parse(reader: &mut Reader) -> Result<Self> {
        let src = reader.source;
        let mut keywords = Default::default();
        let mut detail = Default::default();
        let mut detail_blend_factory = Default::default();
        let mut detail_scale = Default::default();
        let mut detail_blend_mode = Default::default();
        let mut base_texture = Default::default();
        let mut ss_bump = Default::default();
        let mut bump_map = Default::default();

        let mut rest: HashMap<String, Entry> = Default::default();

        loop {
            let (span, key, value) = match event(reader, "item or group end")? {
                Event::GroupEnd(_) => break,
                Event::GroupStart(start) => (
                    start.span,
                    start.name.to_ascii_lowercase(),
                    Entry::Table(Table::load(reader)?),
                ),
                Event::Entry(entry) => (
                    entry.span,
                    entry.key.into_content().to_ascii_lowercase(),
                    entry.value.into(),
                ),
            };
            match key.as_str() {
                "%keywords" => {
                    keywords = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "keywords", span.into(), err)
                    })?;
                }
                "$detail" => {
                    detail = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "detail", span.into(), err)
                    })?;
                }
                "$detailblendfactor" => {
                    detail_blend_factory = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "detail_blend_factory", span.into(), err)
                    })?;
                }
                "$detailscale" => {
                    detail_scale = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "detail_scale", span.into(), err)
                    })?;
                }
                "$detailblendmode" => {
                    detail_blend_mode = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "detail_blend_mode", span.into(), err)
                    })?;
                }
                "$basetexture" => {
                    base_texture = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "base_texture", span.into(), err)
                    })?;
                }
                "$ssbump" => {
                    ss_bump = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "ss_bump", span.into(), err)
                    })?;
                }
                "$bumpmap" => {
                    bump_map = FromEntry::from_entry(value).map_err(|err| {
                        ParseValueError::new(src.into(), "bump_map", span.into(), err)
                    })?;
                }
                _ => {
                    rest.insert(key, value);
                }
            }
        }

        Ok(LightMappedGenericMaterial {
            keywords,
            detail,
            detail_blend_factory,
            detail_scale,
            detail_blend_mode,
            base_texture,
            ss_bump,
            bump_map,
            rest,
        })
    }
}

#[derive(Debug, Clone)]
pub struct OtherMaterial {
    pub name: String,
    pub values: HashMap<String, Entry>,
}

impl OtherMaterial {
    fn parse(name: String, reader: &mut Reader) -> Result<Self> {
        let mut values: HashMap<String, _> = HashMap::new();

        loop {
            let (key, value) = match event(reader, "item or group end")? {
                Event::GroupEnd(_) => break,
                Event::GroupStart(start) => (
                    start.name.to_ascii_lowercase(),
                    Entry::Table(Table::load(reader)?),
                ),
                Event::Entry(entry) => (
                    entry.key.into_content().to_ascii_lowercase(),
                    entry.value.into(),
                ),
            };
            values.insert(key, value);
        }

        Ok(OtherMaterial { name, values })
    }
}

fn event<'a>(reader: &mut Reader<'a>, expected: &'static str) -> Result<Event<'a>> {
    Ok(reader
        .next()
        .ok_or_else(|| EofError::new(reader.source.into(), expected))??)
}

fn expect_event<'a, E: 'a>(reader: &mut Reader<'a>, expected: &'static str) -> Result<E>
where
    E: TryFrom<Event<'a>, Error = VdfError>,
{
    let event = event(reader, expected)?;
    E::try_from(event).map_err(|e| {
        match e {
            VdfError::WrongEntryType(e) => e.with_source(reader.source.into()).into(),
            e => e,
        }
        .into()
    })
}
