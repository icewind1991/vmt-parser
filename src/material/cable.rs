use super::deserialize_path;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CableMaterial {
    /// Defines an albedo texture.
    #[serde(rename = "$basetexture", deserialize_with = "deserialize_path")]
    pub base_texture: String,
    /// Specifies a texture that will provide three-dimensional lighting information for a material.
    #[serde(rename = "$bumpmap", default, deserialize_with = "deserialize_path")]
    pub bump_map: Option<String>,

    /// Use computed vertex colors.
    #[serde(rename = "$vertexcolor", default)]
    pub vertex_color: bool,

    /// Minimum amount of light received
    #[serde(rename = "$minlight", default = "default_min_light")]
    pub min_light: f32,
    /// Maximum amount of light received
    #[serde(rename = "$maxlight", default = "default_max_light")]
    pub max_light: f32,

    /// Disables backface culling.
    #[serde(rename = "$nocull", default)]
    pub no_cull: bool,
}

fn default_min_light() -> f32 {
    0.1
}
fn default_max_light() -> f32 {
    0.3
}
