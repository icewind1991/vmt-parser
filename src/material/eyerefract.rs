use super::deserialize_path;
use crate::{default_scale, Vec3};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EyeRefractMaterial {
    /// Usually referred to as a "sludge-layer", acts as a layer on top of the surface of the $AboveWater Material.
    #[serde(rename = "$iris", deserialize_with = "deserialize_path")]
    pub iris: String,
    /// A texture to specify the shape of the cornea, similar to a normal map. Influences lighting and dilation. The red and green channels are used for the normal mapping, the blue channel is a mask for parallax mapping (straight multiply), and the alpha channel seems to be a multiplier for lighting. Because the $iris is warped in various ways the normals will not match 1-to-1 with the base texture.
    #[serde(rename = "$corneatexture", deserialize_with = "deserialize_path")]
    pub cornea_texture: String,
    /// Strength of the $corneatexture.
    #[serde(rename = "$corneabumpstrength", default = "default_scale")]
    pub cornea_bump_strength: f32,
    /// How much the viewing angle should influence the positioning of the eye
    #[serde(rename = "$parallaxstrength", default = "default_parallax_strength")]
    pub parallax_strength: f32,
    /// Dilates the pupil using the cornea texture to determine the shape of dilation. Default 0.5.
    #[serde(rename = "$dilation", default = "default_dilation")]
    pub dilation: f32,

    /// 1-dimensional texture which remaps lighting colors.
    #[serde(rename = "$lightwarptexture", deserialize_with = "deserialize_path")]
    pub light_warp_texture: String,
    /// Enables cubemap reflections. This shader has a specific cubemap made for it, engine/eye-reflection-cubemap-.vtf, but others can be used, including env_cubemap.
    #[serde(rename = "$envmap", deserialize_with = "deserialize_path")]
    pub env_map: String,
    /// The opacity of the cubemap reflection. Does not affect the eye glint. Default 0.5.
    #[serde(rename = "$glossiness", default = "default_dilation")]
    pub glossiness: f32,

    /// An ambient occlusion texture overlaid onto the entire eye
    #[serde(rename = "$ambientoccltexture", deserialize_with = "deserialize_path")]
    pub ambient_occlusion_texture: String,
    /// Tints the $ambientoccltexture
    #[serde(rename = "$ambientocclcolor", default = "default_occl_color")]
    pub ambient_occlusion_color: Vec3,
    /// Strength of the dynamic ambient occlusion.
    #[serde(rename = "$ambientocclusion", default = "default_scale")]
    pub ambiento_cclusion: f32,

    /// Enables half-lambertian lighting
    #[serde(rename = "$halflambert", default)]
    pub half_lambert: bool,
    /// Enables sphere raytracing. Each pixel is raytraced to allow sharper angles to look more accurate.
    #[serde(rename = "$raytracesphere", default)]
    pub ray_trace_sphere: bool,
    /// Requires $raytracesphere 1. Causes pixels which don't hit the raytraced sphere to be transparent, instead of rendering the "non-raytraced" eye behind it.
    #[serde(rename = "$spheretexkillcombo", default)]
    pub sphere_tex_kill_combo: bool,
    /// Requires $raytracesphere 1. Radius of the eyeball. Should be the diameter of the eyeball divided by 2
    #[serde(rename = "$eyeballradius", default = "default_dilation")]
    pub eye_ball_radius: f32,
}

fn default_parallax_strength() -> f32 {
    0.25
}

fn default_dilation() -> f32 {
    0.5
}

fn default_occl_color() -> Vec3 {
    Vec3([0.33; 3])
}
