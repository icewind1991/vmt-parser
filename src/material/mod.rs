mod lightmappedgeneric;
mod unlitgeneric;
mod water;
mod worldvertextransition;

pub use lightmappedgeneric::LightMappedGenericMaterial;
use serde::{Deserialize, Serialize};
pub use unlitgeneric::UnlitGenericMaterial;
use vdf_reader::entry::{Entry, Table};
use vdf_reader::error::UnknownError;
use vdf_reader::{from_entry, VdfError};
pub use water::WaterMaterial;
pub use worldvertextransition::WorldVertexTransitionMaterial;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Material {
    #[serde(rename = "lightmappedgeneric")]
    LightMappedGeneric(LightMappedGenericMaterial),
    #[serde(rename = "unlitgeneric")]
    UnlitGeneric(UnlitGenericMaterial),
    #[serde(rename = "water")]
    Water(WaterMaterial),
    #[serde(rename = "worldvertextransition")]
    WorldVertexTransition(WorldVertexTransitionMaterial),
    #[serde(rename = "patch")]
    Patch(PatchMaterial),
}

impl Material {
    /// If the material is a patch, apply it to the included material
    pub fn resolve<E, Loader>(self, loader: Loader) -> Result<Material, E>
    where
        Loader: FnOnce(&str) -> Result<String, E>,
        E: From<VdfError>,
    {
        match self {
            Material::Patch(patch) => patch.resolve(loader),
            mat => Ok(mat),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PatchMaterial {
    include: String,
    replace: Table,
}

impl PatchMaterial {
    /// Load the included material and apply the patch
    pub fn resolve<E, Loader>(&self, loader: Loader) -> Result<Material, E>
    where
        Loader: FnOnce(&str) -> Result<String, E>,
        E: From<VdfError>,
    {
        let base = loader(&self.include)?.to_ascii_lowercase();
        let mut material = Table::load_from_str(&base)?;

        let material_values = match material.iter_mut().next() {
            Some((_, Entry::Table(table))) => table,
            _ => {
                return Err(VdfError::from(UnknownError::from(
                    "included vdf doesn't look like a material",
                ))
                .into())
            }
        };
        for (key, value) in self.replace.iter() {
            material_values.insert(key.clone(), value.clone());
        }

        Ok(from_entry(Entry::Table(material))?)
    }
}
