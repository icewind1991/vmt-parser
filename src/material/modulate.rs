use super::deserialize_path;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ModulateMaterial {
    /// Multiplies the color behind it with this surface's texture.
    #[serde(rename = "$basetexture", deserialize_with = "deserialize_path")]
    pub base_texture: String,
    /// Doubles the modulation, making it appear brighter. 1 enables this, 0 disables. Disabled by default.
    #[serde(rename = "$mod2x", default)]
    pub mod_2x: bool,
    /// Disables backface culling.
    #[serde(rename = "$nocull", default)]
    pub no_cull: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "snake_case")]
pub enum SpriteOrientation {
    ParallelUpright,
    #[default]
    VpParallel,
    Oriented,
    VpParallelOriented,
}
