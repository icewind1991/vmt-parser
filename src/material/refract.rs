use super::deserialize_path;
use crate::{
    default_detail_scale, default_scale, default_scale3, BlendMode, TextureTransform, Vec2, Vec3,
};
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct RefractMaterial {
    /// The pattern of refraction is defined by a normal map (DX9+) or DUDV map (DX8-). May be animated.
    #[serde(rename = "$normalmap", deserialize_with = "deserialize_path")]
    pub normal_map: String,
    /// The pattern of refraction is defined by a normal map (DX9+) or DUDV map (DX8-). May be animated.
    #[serde(rename = "$dudvmap", default, deserialize_with = "deserialize_path")]
    pub du_dv_map: Option<String>,
    /// If a second normal map is specified, it will be blended with the first one.
    #[serde(rename = "$normalmap2", default, deserialize_with = "deserialize_path")]
    pub normal_map2: Option<String>,
    /// Use a texture instead of rendering the view for the source of the distorted pixels.
    #[serde(
        rename = "$basetexture",
        default,
        deserialize_with = "deserialize_path"
    )]
    pub base_texture: Option<String>,
    /// Transforms the bump map texture.
    #[serde(rename = "$bumptransform", default)]
    pub bump_transform: TextureTransform,
    /// Transforms the bump map texture.
    #[serde(rename = "$bumptransform2", default)]
    pub bump_transform2: TextureTransform,

    #[serde(rename = "$refracttint", default = "default_scale3")]
    pub refract_tint: Vec3,
    /// Tints the colour of the refraction either uniformly or per-texel. Can be used in conjunction with $refracttint
    #[serde(
        rename = "$refracttinttexture",
        default,
        deserialize_with = "deserialize_path"
    )]
    pub refract_tint_texture: Option<String>,
    /// Controls the strength of the refraction by multiplying the normal map intensity.
    #[serde(rename = "$refractamount", default = "default_scale")]
    pub refract_amount: f32,
    /// Adds a blur effect. Valid values are 0, 1 and 2 (0 and 1 for DX8-).
    #[serde(rename = "$bluramount", default)]
    pub blur_amount: f32,

    #[serde(rename = "$decalscale", default = "default_detail_scale")]
    /// Fits the detail texture onto the material the given number of times
    pub detail_scale: Vec2,
    /// Controls the amount that the detail texture affects the base texture. The precise use of this depends on the blend factor; in most cases it acts similarly to $alpha. A value of 0 usually makes the detail texture have no effect, whilst a value of 1 applies the full effect.
    #[serde(rename = "$detailblendfactor", default = "default_scale")]
    pub detail_blend_factor: f32,
    /// How to combine the detail material with the albedo.
    #[serde(rename = "$detailblendmode", default)]
    pub detail_blend_mode: BlendMode,
    /// A separate VertexLitGeneric material to that will replace this one if the decal hits a model.
    #[serde(
        rename = "$modelmaterial",
        default,
        deserialize_with = "deserialize_path"
    )]
    pub model_material: Option<String>,
    /// Disables texture filtering.
    #[serde(rename = "$pointsamplemagfilter", default)]
    pub point_sample_mag_filter: bool,
    /// Mitigation for displacement texture stretching.
    #[serde(rename = "$seamless_scale", default = "default_scale")]
    pub seamless_scale: f32,

    /// Scales the opacity of an entire material.
    #[serde(rename = "$alpha", default = "default_scale")]
    pub alpha: f32,
    /// Specifies a mask to use to determine binary opacity.
    #[serde(rename = "$alphatest", default)]
    pub alpha_test: bool,
    /// Specifies a mask to use to determine binary opacity.
    #[serde(rename = "$alphatestreference", default = "default_scale")]
    pub alpha_test_reference: f32,
    /// Vector-like edge filtering.
    #[serde(rename = "$distancealpha", default)]
    pub distance_alpha: bool,
    /// Disables backface culling.
    #[serde(rename = "$nocull", default)]
    pub no_cull: bool,
    /// Specifies that the material should be partially see-through.
    #[serde(rename = "$translucent", default)]
    pub translucent: bool,

    /// Specifies a texture that will provide three-dimensional lighting information for a material.
    #[serde(rename = "$bumpmap", default, deserialize_with = "deserialize_path")]
    pub bump_map: Option<String>,
    /// Per-texel color modification via a warp texture.
    #[serde(
        rename = "$lightwarptexture",
        default,
        deserialize_with = "deserialize_path"
    )]
    pub light_wrap_texture: Option<String>,
    /// Determines whether the surface is self-illuminated independent of environment lighting.
    #[serde(rename = "$selfillum", default)]
    pub self_illum: bool,
    /// Flags the $bumpmap as being a self-shadowing bumpmap.
    #[serde(rename = "$ssbump", default)]
    pub ss_bump: bool,

    /// Specular reflections.
    #[serde(rename = "$envmap", default, deserialize_with = "deserialize_path")]
    pub env_map: Option<String>,
    #[serde(rename = "$envmaptint", default = "default_scale3")]
    pub env_map_tint: Vec3,
}
