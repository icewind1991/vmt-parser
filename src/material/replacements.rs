use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use vdf_reader::entry::Table;

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReplacementsMaterial {
    pub templates: HashMap<String, ReplacementTemplate>,
    pub patterns: HashMap<String, ReplacementPattern>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReplacementTemplate(pub HashMap<String, Table>);

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReplacementPattern {
    pub template: String,
}
