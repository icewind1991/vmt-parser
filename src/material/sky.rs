use super::deserialize_path;
use crate::TextureTransform;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SkyMaterial {
    /// Defines an albedo texture.
    #[serde(rename = "$basetexture", deserialize_with = "deserialize_path")]
    pub base_texture: String,
    /// Defines an albedo texture.
    #[serde(
        rename = "$hdrbasetexture",
        default,
        deserialize_with = "deserialize_path"
    )]
    pub hdr_base_texture: Option<String>,
    /// Links the surface to a set of physical properties.
    #[serde(rename = "$surfaceprop", default)]
    pub surface_prop: Option<String>,
    /// Transforms the texture before use in the material. This does not affect lightmaps on the surface.
    #[serde(rename = "$basetexturetransform", default)]
    pub base_texture_transform: TextureTransform,

    /// Ignore z filtering
    #[serde(rename = "$ignorez", default)]
    pub ignore_z: bool,

    /// Prevents fog from overdrawing a material.
    #[serde(rename = "$nofog", default)]
    pub no_fog: bool,

    #[serde(rename = "$nomip", default)]
    pub no_mip: bool,
}
