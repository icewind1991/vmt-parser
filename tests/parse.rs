use miette::{GraphicalReportHandler, GraphicalTheme};
use std::fs::read_to_string;
use test_case::test_case;
use thiserror::Error;
use vmt_parser::from_str;
use vmt_parser::material::Material;

#[derive(Debug, Error)]
enum LoaderError {
    #[error(transparent)]
    Io(#[from] std::io::Error),
    #[error(transparent)]
    Vdf(#[from] vdf_reader::VdfError),
}

#[test_case("tests/data/concretefloor003.vmt")]
#[test_case("tests/data/mvm_backpack.vmt")]
#[test_case("tests/data/water_murky.vmt")]
#[test_case("tests/data/blendrocktograss002.vmt")]
#[test_case("tests/data/patch.vmt")]
#[test_case("tests/data/handrail128_skin2.vmt")]
#[test_case("tests/data/shot2_subrect.vmt")]
fn test_serde(path: &str) {
    let raw = read_to_string(path).unwrap();
    match from_str(&raw) {
        Ok(result) => {
            insta::assert_ron_snapshot!(path, result);
            if let Material::Patch(patch) = result {
                let patched = patch
                    .resolve(|path| read_to_string(path).map_err(LoaderError::from))
                    .unwrap();
                insta::assert_ron_snapshot!(format!("{}_resolved", path), patched);
            }
        }
        Err(e) => {
            let handler = GraphicalReportHandler::new_themed(GraphicalTheme::unicode_nocolor());
            let mut out = String::new();
            handler.render_report(&mut out, &e).unwrap();
            insta::assert_snapshot!(path, out)
        }
    }
}
