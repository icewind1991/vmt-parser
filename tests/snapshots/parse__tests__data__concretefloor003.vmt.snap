---
source: tests/parse.rs
expression: result
---
lightmappedgeneric(LightMappedGenericMaterial(
  r#$basetexture: "cp_mountainlab/concrete/concretefloor003",
  r#$decal: false,
  r#$detail: Some("overlays/detail001"),
  r#$surfaceprop: None,
  r#$basetexturetransform: "center 0.5 0.5 scale 1 1 rotate 0 translate 0 0",
  r#$color: Vec3((1.0, 1.0, 1.0)),
  r#$decalscale: 1.0,
  r#$decalscale: Vec2((4.0, 4.0)),
  r#$detailblendfactor: 1.0,
  r#$detailblendmode: 0,
  r#$modelmaterial: None,
  r#$pointsamplemagfilter: false,
  r#$seamless_scale: 1.0,
  r#$alpha: 1.0,
  r#$alphatest: false,
  r#$alphatestreference: 1.0,
  r#$distancealpha: false,
  r#$nocull: false,
  r#$translucent: false,
  r#$bumpmap: Some("concrete/concretefloor007b_height-ssbump"),
  r#$lightwarptexture: None,
  r#$selfillum: false,
  r#$ssbump: true,
  r#$envmap: None,
  r#$phong: false,
  r#$nofog: false,
  r#$ignorez: false,
))
